docs/content/posts/2023-04-12-howto-migrate-to-datastream.md

```diff
@@ -0,0 +1,190 @@
+after the call to _env.createInput()_.
+
+**
+Datasets**: [DataSet<T>](https://nightlies.apache.org/flink/flink-docs-release-1.12/api/java/org/apache/flink/api/java/DataSet.html)
+are
+now [DataStream<T>](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/datastream/DataStream.html)
+and subclasses.
+
+### [Migrating the join operation](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L131-L137)
+
+The DataStream join operator does not yet support aggregations in batch mode (
+see [FLINK-22587](https://issues.apache.org/jira/browse/FLINK-22587) for details). Basically, the
+problem is with the trigger of the
+default [GlobalWindow](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/windowing/windows/GlobalWindow.html)
+which never fires so the records are never output. We will workaround this problem by applying a
+custom [EndOfStream](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L254-L295)
+window. It is a window assigner that assigns all the records to a
+single [TimeWindow](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/windowing/windows/TimeWindow.html)
+. So, like for the GlobalWindow, all the records are assigned to the same window except that this
+window's trigger is based on the end of the window (which is set to _Long.MAX_VALUE_). As we are on
+a bounded source, at some point the watermark will advance to +INFINITY (Long.MAX_VALUE) and will
+thus cross the end of the time window and consequently fire the trigger and output the records.
+
+Now that we have a working triggering, we just need to call a standard join with the  _Row::join_
+function.
+
+### [Migrating the group by and reduce (sum) operations](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L147-L170)
+
+DataStream API has no
+more [groupBy()](https://nightlies.apache.org/flink/flink-docs-release-1.12/api/java/org/apache/flink/api/java/DataSet.html#groupBy-org.apache.flink.api.java.functions.KeySelector-)
+method, we now use
+the [keyBy()](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/datastream/DataStream.html#keyBy-org.apache.flink.api.java.functions.KeySelector-)
+method. An aggregation downstream will be applied on elements with the same key exactly as
+a [GroupReduceFunction](https://nightlies.apache.org/flink/flink-docs-release-1.12/api/java/org/apache/flink/api/common/functions/GroupReduceFunction.html)
+would have done on a DataSet unless it will not materialize the collection of data. The summing
```

Comment:
> unless it will not materialize the collection of data

Not sure what this is referring to.

Reply:
Wording issue. I meant "except  it will not materialize the collection of data".
Correct me if I'm wrong but DataSet#groupBy() shuffles the data per key and materializes the data belonging to the same key in memory (in a collection) whereas DataStream#keyBy() doesn't do this materialization.

Reply:
> DataSet#groupBy() [...] materializes the data belonging to the same key in memory

That depends on what you're doing afterwards.
A reduce doesn't materialize everything and works similarly to a streaming reduce.

For a join, one/both sides are materialized in memory and/or spilled to disk on the receiver side.
Group Reduce materializes all data on the sender side for sorting purposes.